netty/src/test/java/io/grpc/netty/NettyClientStreamTest.java

```diff
@@ -369,6 +375,34 @@ public void setHttp2StreamShouldNotifyReady() {
     assertTrue(stream.isReady());
   }
 
+  @Test
+  public void removeUserAgentFromApplicationHeaders() {
+    Metadata metadata = new Metadata();
+    metadata.put(GrpcUtil.USER_AGENT_KEY, "bad agent");
+    listener = mock(ClientStreamListener.class);
+    Mockito.reset(writeQueue);
+    when(writeQueue.enqueue(any(), any(boolean.class))).thenReturn(future);
+    ListMultimap<CharSequence, CharSequence> expected = ImmutableListMultimap.copyOf(
+        new DefaultHttp2Headers()
+            .authority(new AsciiString("localhost"))
+            .path(new AsciiString("/" + methodDescriptor.getFullMethodName()))
+            .method(Utils.HTTP_METHOD)
+            .scheme(new AsciiString("http"))
+            .set(Utils.TE_HEADER, Utils.TE_TRAILERS)
+            .set(Utils.CONTENT_TYPE_HEADER, Utils.CONTENT_TYPE_GRPC)
+            .set(Utils.USER_AGENT, AsciiString.of("good agent")));
+
+    stream = new NettyClientStreamImpl(methodDescriptor, new Metadata(), channel, handler,
+        DEFAULT_MAX_MESSAGE_SIZE, AsciiString.of("localhost"), AsciiString.of("http"),
+        AsciiString.of("good agent"));
+    stream.start(listener);
+
+    ArgumentCaptor<CreateStreamCommand> cmdCap = ArgumentCaptor.forClass(CreateStreamCommand.class);
+    verify(writeQueue).enqueue(cmdCap.capture(), eq(false));
+    assertThat(ImmutableListMultimap.copyOf(cmdCap.getValue().headers()))
+        .containsExactlyEntriesIn(expected);
```

Comment:
Key implies map, which netty headers are not.   They are a loosely ordered table.  I do expect headers to be in the order encoded by GrpcHttp2Headers.